File: adminapp/src/main/scala/org/openurp/edu/laboratory/teacher/web/action/ApplyAction.scala
package org.openurp.edu.laboratory.teacher.web.action

import org.beangle.webmvc.entity.action.RestfulAction
import org.beangle.data.dao.OqlBuilder
import org.openurp.base.model.Semester
import org.openurp.platform.api.security.Securities
import org.openurp.edu.base.model.Teacher
import org.openurp.edu.lesson.model.Lesson
import org.openurp.base.model.Building
import org.openurp.edu.laboratory.model.Software
import org.beangle.webmvc.api.view.View
import org.beangle.commons.lang.Numbers
import org.beangle.webmvc.api.annotation.param
import org.openurp.base.model.Building
import org.openurp.base.model.Semester
import org.openurp.edu.laboratory.model.Software
import org.openurp.edu.lesson.model.Lesson
import org.openurp.edu.laboratory.model.ExprProgram
import org.openurp.edu.laboratory.model.LabRoomApply
import org.openurp.base.model.Building
import org.openurp.base.model.Semester
import org.openurp.edu.laboratory.model.ExprProgram
import org.openurp.edu.laboratory.model.Software
import org.openurp.edu.lesson.model.Lesson

/**
 * 实验室申请
 */
class ApplyAction extends RestfulAction[LabRoomApply] {

  def getTeacher(): Teacher = {
    val teachers = entityDao.search(OqlBuilder.from(classOf[Teacher], "s").where("s.user.code=:code", Securities.user))
    if (teachers.isEmpty) {
      throw new RuntimeException("Cannot find teachers with code " + Securities.user)
    } else {
      teachers.head
    }
  }

  override def indexSetting(): Unit = {
    val semesters = entityDao.search(OqlBuilder.from(classOf[Semester], "semester").where("semester.calendar.school.id=:id", 16404))
    put("semesters", semesters)
    val semesterQuery = OqlBuilder.from(classOf[Semester], "semester").where(":now between semester.beginOn and semester.endOn", new java.util.Date())
    semesterQuery.where("semester.calendar.school.id=:id", 16404)
    put("currentSemester", entityDao.search(semesterQuery).head)
    super.indexSetting()
  }
  /**
   * 此处应该找到用户的申请
   */
  override def search(): String = {
    val semesterQuery = OqlBuilder.from(classOf[Semester], "semester").where(":now between semester.beginOn and semester.endOn", new java.util.Date())
    semesterQuery.where("semester.calendar.school.id=:id", 16404)
    val semesterId = getInt("semester.id").getOrElse(entityDao.search(semesterQuery).head.id)
    val semester = entityDao.get(classOf[Semester], semesterId);
    val teacher = getTeacher()
    if (teacher == null) { forward("error.teacher.teaNo.needed") }
    // 得到该教师的所有教学任务
    val lessonQuery = OqlBuilder.from(classOf[Lesson], "lesson");
    lessonQuery.limit(getPageLimit)
    lessonQuery.where("lesson.semester.id=:semesterId", semesterId);
    if (teacher != null) {
      lessonQuery.join("lesson.teachers", "teacher")
      lessonQuery.where("teacher =:teacher", teacher);
    }
    val lessons = entityDao.search(lessonQuery);
    put("lessons", lessons);
    forward();
  }

  def add(): String = {
    val lessonId = longId("lesson")
    val lesson = entityDao.get(classOf[Lesson], lessonId)
    var apply = entityDao.search(OqlBuilder.from(classOf[LabRoomApply], "apply").where("apply.lesson=:leson", lesson))
    if (apply.isEmpty) {
      val application = new LabRoomApply()
      put("application", application)
    } else put("apply", apply.head)
    put("lesson", lesson)
    val buildings = entityDao.search(OqlBuilder.from(classOf[Building], "building").where("building.school.id=:id", 16404))
    put("buildings", buildings)
    val softwares = entityDao.getAll(classOf[Software])
    put("softwares", softwares)
    forward()
  }

  def save(@param("apply.id") id: String): View = {
    val applyId = Numbers.toLong(id)
    val application = entityDao.get(classOf[LabRoomApply], applyId)
    val softwares = getAll("fake.softwares", classOf[Int])
    application.softwares.clear()
    application.softwares ++= entityDao.find(classOf[Software], softwares)
    super.saveAndRedirect(application)
  }
  
  def addApply(): String = {
    val labId = longId("lab")
    val lab = entityDao.get(classOf[ExprProgram], labId)
    put("lab", lab)
    val lesson = entityDao.get(classOf[Lesson], lab.lesson.id)
    put("lesson", lesson)
    val applies = entityDao.search(OqlBuilder.from(classOf[LabRoomApply], "apply").where("apply.lab=:lab", lab))
    if (applies.isEmpty) { put("apply", null) }
    else put("apply", applies.head)
    val buildings = entityDao.search(OqlBuilder.from(classOf[Building], "building").where("building.school.id=:id", 16404))
    put("buildings", buildings)
    val softwares = entityDao.getAll(classOf[Software])
    put("softwares", softwares)
    forward()
  }

  def saveApply(@param("lab.id") id: String): View = {
    val labId = Numbers.toLong(id)
    val lab = entityDao.get(classOf[LabRoomApply], labId)
    val apply = new LabRoomApply()
//    apply.lab = lab
    val softwares = getAll("fake.softwares", classOf[Int])
    apply.softwares.clear()
    apply.softwares ++= entityDao.find(classOf[Software], softwares)
//    val building = entityDao.get(classOf[Building], intId("apply.labBuilding"))
//    apply.labBuilding = building
    entityDao.saveOrUpdate(apply)

    redirect("search", "info.save.success")
  }
}

File: adminapp/src/main/java/org/openurp/edu/laboratory/teacher/web/action/ApplyAction.java
package org.openurp.edu.laboratory.teacher.web.action;

import java.util.Collection;
import java.util.List;
import java.util.Set;

import org.beangle.commons.collection.CollectUtils;
import org.beangle.commons.dao.query.builder.OqlBuilder;
import org.beangle.commons.lang.Strings;
import org.beangle.commons.lang.time.WeekState;
import org.beangle.commons.lang.time.WeekTime;
import org.beangle.commons.lang.time.WeekTimes;
import org.beangle.struts2.convention.route.Action;
import org.openurp.base.model.Semester;
import org.openurp.base.model.TimeSetting;
import org.openurp.base.util.WeekStates;
import org.openurp.base.util.WeekTimeBuilder;
import org.openurp.edu.base.model.AuditState;
import org.openurp.edu.base.model.Classroom;
import org.openurp.edu.base.model.Teacher;
import org.openurp.edu.laboratory.model.ExprProgram;
import org.openurp.edu.laboratory.model.LabRoomApply;
import org.openurp.edu.laboratory.model.Software;
import org.openurp.edu.lesson.model.CourseActivity;
import org.openurp.edu.lesson.model.Lesson;
import org.openurp.edu.lesson.util.CourseActivityDigestor;
import org.openurp.edu.web.action.TeacherProjectSupport;

public class ApplyAction extends TeacherProjectSupport {

  @Override
  public String innerIndex() {

    return null;
  }

  public String edit() {
    return forward();
  }

  private Set<CourseActivity> getApplyableActivities(Lesson lesson) {
    if (true) return lesson.getCourseSchedule().getActivities();

    Set<CourseActivity> activities = lesson.getCourseSchedule().getActivities();
    Set<CourseActivity> applyableActivities = CollectUtils.newHashSet();
    for (CourseActivity ca : activities) {
      if (ca.getRooms().isEmpty()) {
        applyableActivities.add(ca);
      } else {
        for (Classroom cl : ca.getRooms()) {
          if (cl.getRoom() == null) {
            applyableActivities.add(ca);
          }
        }
      }
    }
    return applyableActivities;
  }

  private List<WeekTime> toRelativeTimes(Lesson lesson, Collection<CourseActivity> activities) {
    Semester semester = lesson.getSemester();
    List<WeekTime> times = CollectUtils.newArrayList();
    for (CourseActivity ca : activities) {
      WeekTime time = (WeekTime) ca.getTime().clone();
      int weekOffset = WeekTimeBuilder.getOffset(semester, time.getWeekday());
      int reverseOffset = WeekTimeBuilder.getReverseOffset(semester, time.getWeekday());
      time.setStartOn(WeekTimeBuilder.getStartOn(semester, time.getWeekday()));
      if (time.getStartOn().getYear() == semester.getBeginOn().getYear()) {
        time.setWeekstate(new WeekState(time.getWeekstate().value >> weekOffset));
      } else {
        time.setWeekstate(new WeekState(time.getWeekstate().value << reverseOffset));
      }
      times.add(time);
    }
    times = WeekTimes.mergeTimes(times);
    return times;
  }

  public String applyByLesson() {
    Teacher teacher = getLoginTeacher();
    Long applyId = getLong("apply.id");
    Lesson lesson = null;
    LabRoomApply apply = null;
    if (null != applyId) {
      apply = entityDao.get(LabRoomApply.class, applyId);
      lesson = apply.getLesson();
    } else {
      Long lessonId = getLong("lesson.id");
      lesson = entityDao.get(Lesson.class, lessonId);
      apply = new LabRoomApply();
      apply.setLesson(lesson);

      OqlBuilder<ExprProgram> pbuilder = OqlBuilder.from(ExprProgram.class, "p");
      pbuilder.where("p.lesson=:lesson", lesson);
      List<ExprProgram> programs = entityDao.search(pbuilder);
      apply.getSoftwares().addAll(programs.get(0).getSoftwares());
      apply.setTel(getTelphone(teacher, lesson.getSemester()));

    }
    if (!lesson.getTeachers().contains(teacher)) { return forwardError("不是你的课程"); }

    TimeSetting timeSetting = timeSettingService.getClosestTimeSetting(lesson.getProject(),
        lesson.getSemester(), lesson.getCampus());
    put("timeSetting", timeSetting);

    Set<CourseActivity> applyableActivities = getApplyableActivities(lesson);
    put("applyableActivityText", CourseActivityDigestor.getInstance().digest(null, timeSetting,
        applyableActivities, ":day :units :weeks :room"));
    if (apply.isTransient()) {
      for (CourseActivity ca : applyableActivities) {
        if (ca.getRooms().isEmpty()) {
          apply.getTimes().add(ca.getTime());
        } else {
          for (Classroom cl : ca.getRooms()) {
            if (cl.getRoom() == null) {
              apply.getTimes().add(ca.getTime());
            }
          }
        }
      }
    }

    Set<java.util.Date> existedDates = CollectUtils.newHashSet();
    for (WeekTime s : apply.getTimes()) {
      existedDates.addAll(s.getDates());
    }
    put("existedDates", existedDates);
    List<WeekTime> times = toRelativeTimes(lesson, applyableActivities);
    put("times", times);

    OqlBuilder<Software> sbuilder = OqlBuilder.from(Software.class, "s");
    // sbuilder.where("s.project=:project", lesson.getProject());
    sbuilder.orderBy("s.name");
    put("softwares", entityDao.search(sbuilder));

    put("apply", apply);
    return forward("lessonForm");
  }

  public String saveByLesson() {
    Teacher teacher = getLoginTeacher();
    Long applyId = getLong("apply.id");
    Lesson lesson = null;
    LabRoomApply apply = null;
    if (null != applyId) {
      apply = entityDao.get(LabRoomApply.class, applyId);
      lesson = apply.getLesson();
    } else {
      Long lessonId = getLong("lesson.id");
      lesson = entityDao.get(Lesson.class, lessonId);
      apply = new LabRoomApply();
    }
    populate(apply, "apply");
    apply.setLesson(lesson);
    apply.setProject(lesson.getProject());
    apply.setSemester(lesson.getSemester());
    apply.setBorrower(teacher.getUser());
    apply.setActivity(lesson.getNo() + " " + lesson.getCourse().getName());
    apply.setAudience(lesson.getTeachClass().getName());
    apply.setAttendance(lesson.getTeachClass().getStdCount());
    apply.setState(AuditState.SUBMITTED);
    apply.setUpdatedAt(new java.util.Date());
    apply.getTimes().clear();
    if (!lesson.getTeachers().contains(teacher)) { return forwardError("不是你的课程"); }
    Set<CourseActivity> applyableActivities = getApplyableActivities(lesson);
    List<WeekTime> times = toRelativeTimes(lesson, applyableActivities);
    List<WeekTime> selected = CollectUtils.newArrayList();
    for (WeekTime time : times) {
      String name = "weekId_" + time.getWeekday().getId() + "_" + time.getBeginAt() + "_" + time.getEndAt();
      String[] weeks = getAll(name, String.class);
      if (null != weeks && weeks.length > 0) {
        time.setWeekstate(WeekStates.build(Strings.join(weeks)));
        selected.add(time);
      }
    }
    if (!selected.isEmpty()) {
      for (WeekTime wt : selected) {
        WeekTimeBuilder builder = WeekTimeBuilder.on(lesson.getSemester());
        List<WeekTime> rs = builder.build(wt.getWeekday(), wt.getWeekstate().getWeekList());
        for (WeekTime wtrs : rs) {
          wtrs.setBeginAt(wt.getBeginAt());
          wtrs.setEndAt(wt.getEndAt());
        }
        apply.getTimes().addAll(rs);
      }
    }
    String[] softwareIds = getAll("software.id", String.class);
    apply.getSoftwares().clear();
    if (softwareIds != null && softwareIds.length > 0) {
      apply.getSoftwares().addAll(entityDao.get(Software.class, Strings.transformToInt(softwareIds)));
    }
    if (!apply.getTimes().isEmpty()) {
      entityDao.saveOrUpdate(apply);
    }
    return redirect(new Action(ProgramAction.class, "lessons"), "info.save.success", null);
  }

  private String getTelphone(Teacher teacher, Semester semester) {
    OqlBuilder<String> builder = OqlBuilder.from(LabRoomApply.class.getName(), "apply");
    builder.where("apply.project =:project", teacher.getProject());
    builder.where("apply.semester = :semester", semester);
    builder.where(":teacher in elements(apply.lesson.teachers)", teacher);
    builder.select("distinct apply.tel");
    List<String> tels = entityDao.search(builder);
    if (tels.isEmpty()) return null;
    else return tels.get(0);

  }

}
